### PR TITLE
New secret

### PR DESCRIPTION
**Pull request by Eabramov84**

### CODE REVIEW

File: iamserver/iamserver.go [hunk 1]
```diff
@@ -97,8 +98,9 @@ type IdentHandler interface {
 
 // PermissionHandler interface.
 type PermissionHandler interface {
-	RegisterService(serviceID string, funcServerPermissions map[string]map[string]string) (secret string, err error)
-	UnregisterService(serviceID string)
+	RegisterInstance(
+		instance cloudprotocol.InstanceIdent, permissions map[string]map[string]string) (secret string, err error)
+	UnregisterService(instance cloudprotocol.InstanceIdent)
```
Comment (al1img) [inline iamserver/iamserver.go:103]: UnregisterInstance

File: permhandler/permhandler.go [hunk 2]
```diff
@@ -86,30 +90,40 @@ func (handler *Handler) RegisterService(
...
+		permissions: permissions, instancesCount: 1,
+	}
 
 	return string(newSecret), nil
 }
 
-// UnregisterService deletes service with permissions from cache.
-func (handler *Handler) UnregisterService(serviceID string) {
+// UnregisterService deletes service instance with permissions from cache.
+func (handler *Handler) UnregisterService(instance cloudprotocol.InstanceIdent) {
```
Comment (al1img) [inline permhandler/permhandler.go:102]: UnregisterInstance

File: iamserver/iamserver.go [hunk 3]
```diff
@@ -375,19 +381,20 @@ func (server *Server) SubscribeSubjectsChanged(message *empty.Empty,
...
-	ctx context.Context, req *pb.RegisterServiceRequest) (rsp *pb.RegisterServiceResponse, err error) {
-	rsp = &pb.RegisterServiceResponse{}
+// RegisterInstance registers new service and creates secret.
+func (server *Server) RegisterInstance(
+	ctx context.Context, req *pb.RegisterInstanceRequest,
+) (*pb.RegisterInstanceResponse, error) {
+	rsp := &pb.RegisterInstanceResponse{}
 
-	log.WithField("serviceID", req.ServiceId).Debug("Process register service")
+	log.WithField("serviceID", req.Instance.ServiceId).Debug("Process register service")
```
Comment (al1img) [inline iamserver/iamserver.go:390]: Process register instance, add also subjectID and instance to the log fields.

File: iamserver/iamserver.go [hunk 4]
```diff
@@ -399,21 +406,19 @@ func (server *Server) RegisterService(
...
 
-// UnregisterService unregisters service.
-func (server *Server) UnregisterService(
-	ctx context.Context, req *pb.UnregisterServiceRequest) (rsp *empty.Empty, err error) {
-	rsp = &empty.Empty{}
-
-	log.WithField("serviceID", req.ServiceId).Debug("Process unregister service")
+// UnregisterInstance unregisters service.
+func (server *Server) UnregisterInstance(ctx context.Context, req *pb.UnregisterInstanceRequest) (*empty.Empty, error) {
+	log.WithField("serviceID", req.Instance.ServiceId).Debug("Process unregister service")
```
Comment (al1img) [inline iamserver/iamserver.go:411]: Process unregister instance, add subject ID and instance to the log field.

File: iamserver/iamserver_test.go [hunk 5]
```diff
@@ -439,187 +446,66 @@ func TestSetUsers(t *testing.T) {
...
+			"vis": {"*": "rw", "test": "r"},
+		}
+		funServerID     = "vis"
+		setPBPermission = &pb.Permissions{Permissions: map[string]string{"*": "rw", "test": "r"}}
+	)
 
-	resp, err := client.pbProtected.RegisterService(ctx, req)
-	if err != nil {
-		t.Fatalf("Can't send request: %s", err)
+	req := &pb.RegisterInstanceRequest{
```
Comment (al1img) [inline iamserver/iamserver_test.go:478]: Add subjectId and instance as well.

File: iamserver/iamserver_test.go [hunk 6]
```diff
@@ -439,187 +446,66 @@ func TestSetUsers(t *testing.T) {
...
 	}
 
-	req = &pb.RegisterServiceRequest{ServiceId: "serviceID2", Permissions: map[string]*pb.Permissions{"vis": permission}}
-
-	resp, err = client.pbProtected.RegisterService(ctx, req)
-	if err != nil {
-		t.Fatalf("Can't send request: %s", err)
+	if receivesPermission, ok := permHandler.permissions[resp.Secret]; ok {
+		if !reflect.DeepEqual(receivesPermission, expectedPermissions) {
+			t.Errorf("Incorrect requested permissions")
```
Comment (al1img) [inline iamserver/iamserver_test.go:490]: Error

File: iamserver/iamserver_test.go [hunk 7]
```diff
@@ -439,187 +446,66 @@ func TestSetUsers(t *testing.T) {
...
+		}
+	} else {
+		t.Error("Permission doesn't received")
```
Comment (al1img) [inline iamserver/iamserver_test.go:493]: is not received

File: iamserver/iamserver_test.go [hunk 8]
```diff
@@ -628,41 +514,29 @@ func TestGetPermissionsServerPublic(t *testing.T) {
...
-		t.Fatalf("Can't send request: %s", err)
+	if !proto.Equal(getPBPermissions.Permissions, setPBPermission) {
+		t.Error("Incorrect permission")
 	}
 
-	if !reflect.DeepEqual(perm.Permissions.Permissions, vis.Permissions) {
-		t.Fatalf("Wrong perm: received %v expected %v", perm, vis.Permissions)
+	if _, err := client.pbPublic.GetPermissions(ctx,
+		&pb.PermissionsRequest{Secret: "noSecret", FunctionalServerId: funServerID}); err == nil {
+		t.Errorf("Should be error")
```
Comment (al1img) [inline iamserver/iamserver_test.go:528]: t.Error

File: iamserver/iamserver_test.go [hunk 9]
```diff
@@ -78,6 +79,12 @@ type testIdentHandler struct {
 	subjectsChangedChannel chan []string
 }
 
+type testPermissionHandler struct {
+	permissions      map[string]map[string]map[string]string
+	currentServiceID string
```
Comment (al1img) [inline iamserver/iamserver_test.go:84]: Store whole ident

File: iamserver/iamserver.go [hunk 10]
```diff
@@ -97,8 +98,9 @@ type IdentHandler interface {
 
 // PermissionHandler interface.
 type PermissionHandler interface {
-	RegisterService(serviceID string, funcServerPermissions map[string]map[string]string) (secret string, err error)
-	UnregisterService(serviceID string)
+	RegisterInstance(
+		instance cloudprotocol.InstanceIdent, permissions map[string]map[string]string) (secret string, err error)
+	UnregisterService(instance cloudprotocol.InstanceIdent)
 	GetPermissions(secret, funcServerID string) (serviceID string, permissions map[string]string, err error)
```
Comment (al1img) [inline iamserver/iamserver.go:104]: Should it return instance ident instead of serviceID?

File: iamserver/iamserver_test.go [hunk 11]
```diff
@@ -835,3 +710,35 @@ func (handler *testIdentHandler) GetSubjects() (subjects []string, err error) {
...
+
+	secret = time.Now().String()
+
+	permission.permissions[secret] = permissions
+	permission.currentServiceID = instance.ServiceID
+
+	return secret, nil
+}
+
+func (permission *testPermissionHandler) UnregisterService(instance cloudprotocol.InstanceIdent) {}
```
Comment (al1img) [inline iamserver/iamserver_test.go:729]: Unregister is not checked.

File: iamserver/iamserver_test.go (hunk 6 above)
Comment (Eabramov84) [reply iamserver/iamserver_test.go:490]: fixed

File: permhandler/permhandler.go [hunk 12]
```diff
@@ -157,12 +159,12 @@ func generateSecret() (secret secretKey, err error) {
 	return secretKey(base64.StdEncoding.EncodeToString(b)), nil
 }
 
-func (handler *Handler) findServiceID(serviceID string) (secret string, err error) {
+func (handler *Handler) getSecretForInstance(instance cloudprotocol.InstanceIdent) (string, error) {
 	for key, value := range handler.secrets {
-		if value.serviceID == serviceID {
+		if reflect.DeepEqual(value.instaneIdent, instance) {
```
Comment (MykolaSolyanko) [inline permhandler/permhandler.go:164]: you can compare instances like variables without using reflect.DeepEqual
```
if value.instaneIdent = instance {
```

**Review by MykolaSolyanko — APPROVED** [review]

File: iamserver/iamserver_test.go (hunk 11 above)
Comment (Eabramov84) [reply iamserver/iamserver_test.go]: implemented

File: permhandler/permhandler.go (hunk 12 above)
Comment (Eabramov84) [reply permhandler/permhandler.go:164]: done, thanks

File: iamserver/iamserver_test.go [hunk 13]
```diff
@@ -439,230 +447,105 @@ func TestSetUsers(t *testing.T) {
...
+	req := &pb.RegisterInstanceRequest{
+		Instance:    &pb.InstanceIdent{ServiceId: testServiceID, SubjectId: "s1", Instance: 2},
+		Permissions: map[string]*pb.Permissions{funServerID: setPBPermission},
 	}
 
-	if resp.Secret == "" {
-		t.Fatal("Incorrect secret")
+	resp, err := client.pbProtected.RegisterInstance(ctx, req)
+	if err != nil {
+		t.Fatalf("Can't send request instance: %v", err)
```
Comment (al1img) [inline iamserver/iamserver_test.go:453]: `can't register instance`?

File: iamserver/iamserver_test.go [hunk 14]
```diff
@@ -835,3 +719,38 @@ func (handler *testIdentHandler) GetSubjects() (subjects []string, err error) {
...
+func (permission *testPermissionHandler) UnregisterInstance(instance cloudprotocol.InstanceIdent) {
+	permission.currentUnregisterInstance = instance
+}
+
+func (permission *testPermissionHandler) GetPermissions(
+	secret, funcServerID string,
+) (cloudprotocol.InstanceIdent, map[string]string, error) {
+	allPermissions, ok := permission.permissions[secret]
+	if !ok {
+		return permission.currentInstance, nil, aoserrors.New("No permissions")
```
Comment (al1img) [inline iamserver/iamserver_test.go:704]: error from small letter.

File: iamserver/iamserver_test.go [hunk 15]
```diff
@@ -835,3 +719,38 @@ func (handler *testIdentHandler) GetSubjects() (subjects []string, err error) {
...
+	}
+
+	permissions, ok := allPermissions[funcServerID]
+	if !ok {
+		return permission.currentInstance, nil, aoserrors.New("No permissions")
```
Comment (al1img) [inline iamserver/iamserver_test.go:709]: ditto

**Review by al1img — APPROVED** [review]

**Pull request merged**